src/rewriter.coffee

```diff
@@ -512,6 +513,23 @@ exports.Rewriter = class Rewriter
         last_column:  prevLocationData.last_column
       return 1
 
+  # Add parens around a `do` IIFE followed by a chained `.` so that the
+  # chaining applies to the executed function rather than the function
+  # object (see #3736)
+  addParensToChainedDoIife: ->
+    condition = (token, i) ->
+      @tag(i - 1) is 'OUTDENT'
+    action = (token, i) ->
+      return unless token[0] is '.'
+      @tokens.splice doIndex, 0, generate '(', '(', @tokens[doIndex][2]
+      @tokens.splice i + 1, 0, generate ')', ')', ['', 'implicit parentheses', token[2]]
```

Comment:
@GeoffreyBooth @lydell I'm not too sure what these origin args should be?

Reply:
`origin` is just a reference to whatever token we copied the location data from. So for `(`, that should be the `do` token, and for `)`, that should be whatever token immediately precedes the `)` after it's inserted.

Reply:
@GeoffreyBooth ok updated origin args to reference those tokens (I believe)